bin/test-example

```diff
@@ -0,0 +1,60 @@
+#!/bin/sh
+
+exampledir=$1
+
+examplename=$(basename "$exampledir")
+exercisedir=$(dirname $(dirname "$exampledir"))
+exercisename=$(basename "$exercisedir")
+xhaskell=$(dirname $(dirname "$0"))
+
+buildfolder="${TRAVIS_BUILD_DIR:-$xhaskell}/build/${exercisename}/${examplename}"
+mkdir -p "${buildfolder}"
+cp -rL ${exercisedir}/stack.yaml ${exercisedir}/test ${exampledir}/* "${buildfolder}"
+
+cd $buildfolder
+
+if [ -n "$TRAVIS" ]; then
+  cachedir="$HOME"
+else
+  cachedir="$xhaskell"
+fi
+examplecache="${cachedir}/.foldercache/${exercisename}/${examplename}/.stack-work"
+mkdir -p "$examplecache"
+ln -f -s "$examplecache"
+
+exampletype=$(echo "$examplename" | cut -d- -f1)
+
+runstack () {
+    # SET_RESOLVER passed by .travis.yml - sets --resolver if not current.
+    stack "$@" ${SET_RESOLVER} `# Select the correct resolver. `\
+               --install-ghc   `# Download GHC if not in cache.`\
+               --no-terminal   `# Terminal detection is broken.`\
+               --pedantic      `# Enable -Wall and -Werror.    `
+}
+
+if [ "$exampletype" = "success" ]; then
+    echo "testing ${exampledir} - should succeed"
+    # Implicit exit value: this is last thing in this path,
+    # so the script's exit value = stack's exit value.
+    # If statements are reordered, please preserve this property.
+    runstack "test"
+elif [ "$exampletype" = "fail" ]; then
+    echo "testing ${exampledir} - should build, but fail tests"
+    if ! runstack "build"; then
+        echo "${exampledir} build failed unexpectedly"
+        exit 1
+    fi
+    if runstack "test"; then
+        echo "${exampledir} test succeeded unexpectedly"
+        exit 1
+    fi
+elif [ "$exampletype" = "error" ]; then
+    echo "testing ${exampledir} - should fail to build"
+    if runstack "test" "--no-run-tests"; then
```

Comment:
this excessive quoting is not really necessary. it could be removed, if requested.